lib/Authentication.ts

```diff
@@ -49,6 +52,150 @@ export default class Authentication {
     this.factory = new CryptoFactory(options.cryptoSuites || [new RsaCryptoSuite()]);
   }
 
+  /**
+   * Signs the AuthenticationRequest with the private key of the Requester and returns the signed JWT.
+   * @param request well-formed AuthenticationRequest object
+   * @param responseDid DID of the requester.
+   */
+  public async signAuthenticationRequest (request: AuthenticationRequest): Promise<string> {
+    if (request.response_type !== 'id_token' || request.scope !== 'openid') {
+      throw new Error('Authentication Request not formed correctly');
+    }
+    const requesterDid = request.iss;
+    let key: PrivateKey | undefined;
+    key = this.getKey(requesterDid);
+    if (!key) {
+      throw new Error(`Could not find a key for ${requesterDid}`);
+    }
+
+    const token = new JwsToken(request, this.factory);
+    return token.sign(key);
+  }
+
+  /**
+   * Verifies signature on request and returns AuthenticationRequest.
+   * @param request Authentiation Request as a buffer or string.
+   */
+  public async verifyAuthenticationRequest (request: Buffer | string): Promise<AuthenticationRequest> {
+    // identity function in place of later signature verification
+    let jwsToken: JwsToken;
+    if (request instanceof Buffer) {
+      jwsToken = new JwsToken(request.toString(), this.factory);
+    } else {
+      jwsToken = new JwsToken(request, this.factory);
+    }
+    const keyId = jwsToken.getHeader().kid;
+    const keyDid = DidDocument.getDidFromKeyId(keyId);
+    const results = await this.resolver.resolve(keyDid);
+    const didPublicKey = results.didDocument.getPublicKey(keyId);
+    if (!didPublicKey) {
+      throw new Error('Could not find public key');
+    }
+    const publicKey = this.factory.constructPublicKey(didPublicKey);
+    const content = await jwsToken.verifySignature(publicKey);
+
+    const verifiedRequest: AuthenticationRequest = JSON.parse(content);
+    if (verifiedRequest.iss !== keyDid) {
+      throw new Error('Signing DID does not match issuer');
+    }
+    return verifiedRequest;
+  }
+
+  /**
+   * Given a challenge, forms a signed response using a given DID that expires at expiration, or a default expiration.
+   * @param authRequest Challenge to respond to
+   * @param responseDid The DID to respond with
+   * @param claims Claims that the requester asked for
+   * @param expiration optional expiration datetime of the response
+   */
+  public async formAuthenticationResponse (authRequest: AuthenticationRequest, responseDid: string, claims: any, expiration?: Date): Promise<string> {
+    let key: PrivateKey | undefined;
+    key = this.getKey(responseDid);
+    if (!key) {
+      throw new Error(`Could not find a key for ${responseDid}`);
+    }

```

Comment:
I've seen these lines twice now next to the same function:
```
    if (!key) {
      throw new Error(`Could not find a key for ${responseDid}`);
    }
```
Can this move into the `getKey` function and just always return a `PrivateKey`?